In [47]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.preprocessing import LabelEncoder
from sklearn import tree
from sklearn.linear_model import LogisticRegression

In [18]:
titanic = pd.read_csv("titanic.csv")
columns_to_drop = ['PassengerId', 'Name', 'SibSp', 'Parch', 'Ticket', 'Cabin', 'Embarked']
titanic.drop(columns=columns_to_drop, axis=1, inplace=True)

In [93]:
inputs = titanic.drop('Survived',axis='columns')
target= titanic.Survived
print(inputs)
print("***********************************")
print(target)

     Pclass     Sex   Age     Fare
0         3    male  22.0   7.2500
1         1  female  38.0  71.2833
2         3  female  26.0   7.9250
3         1  female  35.0  53.1000
4         3    male  35.0   8.0500
..      ...     ...   ...      ...
886       2    male  27.0  13.0000
887       1  female  19.0  30.0000
888       3  female   NaN  23.4500
889       1    male  26.0  30.0000
890       3    male  32.0   7.7500

[891 rows x 4 columns]
***********************************
0      0
1      1
2      1
3      1
4      0
      ..
886    0
887    1
888    0
889    1
890    0
Name: Survived, Length: 891, dtype: int64


In [23]:
le_Sex = LabelEncoder()
inputs['Sex'] = le_Sex.fit_transform(inputs['Sex'])
inputs['Age'] = inputs['Age'].fillna(inputs['Age'].mean())
print(inputs)

     Pclass  Sex        Age     Fare
0         3    1  22.000000   7.2500
1         1    0  38.000000  71.2833
2         3    0  26.000000   7.9250
3         1    0  35.000000  53.1000
4         3    1  35.000000   8.0500
..      ...  ...        ...      ...
886       2    1  27.000000  13.0000
887       1    0  19.000000  30.0000
888       3    0  29.699118  23.4500
889       1    1  26.000000  30.0000
890       3    1  32.000000   7.7500

[891 rows x 4 columns]


In [52]:
model1 = RandomForestClassifier(n_estimators=50)
model2= SVC(kernel='rbf')
model3=tree.DecisionTreeClassifier()
model4=LogisticRegression()

In [87]:
# Evaluate models using cross-validation
models = [model1, model2, model3, model4]
model_names = ['RandomForest', 'SVC', 'DecisionTree', 'LogisticRegression']
for model, name in zip(models, model_names):
    scores = cross_val_score(model, inputs, target)
    print(f"{name} Average CV Score: {scores.mean()}")
    print("\n")

RandomForest Average CV Score: 0.8137153976523759


SVC Average CV Score: 0.6701462557278262


DecisionTree Average CV Score: 0.7800954114619296


LogisticRegression Average CV Score: 0.7867553825874082




**Split data for final evaluation on the test set for a final unbiased evaluation**


In [89]:
X_train, X_test, Y_train, Y_test = train_test_split(inputs, target, test_size=0.2, random_state=42)

for model, name in zip(models, model_names):
    model.fit(X_train, Y_train)
    print(f"{name} Test Accuracy: {model.score(X_test,Y_test)}")
    print("\n")


RandomForest Test Accuracy: 0.8156424581005587


SVC Test Accuracy: 0.6536312849162011


DecisionTree Test Accuracy: 0.7541899441340782


LogisticRegression Test Accuracy: 0.7988826815642458




# **Clearly the best model for the prediction of titanic survival as far is Random forest with accuracy ≈ 81%** 